whois-api/src/main/java/net/ripe/db/whois/api/rdap/RdapRelationService.java

```diff
@@ -0,0 +1,198 @@
+package net.ripe.db.whois.api.rdap;
+
+import com.google.common.collect.Sets;
+import net.ripe.db.whois.api.rdap.domain.RelationType;
+import net.ripe.db.whois.common.dao.RpslObjectDao;
+import net.ripe.db.whois.common.domain.CIString;
+import net.ripe.db.whois.common.ip.IpInterval;
+import net.ripe.db.whois.common.ip.Ipv4Resource;
+import net.ripe.db.whois.common.iptree.IpEntry;
+import net.ripe.db.whois.common.iptree.IpTree;
+import net.ripe.db.whois.common.iptree.Ipv4DomainTree;
+import net.ripe.db.whois.common.iptree.Ipv4Tree;
+import net.ripe.db.whois.common.iptree.Ipv6DomainTree;
+import net.ripe.db.whois.common.iptree.Ipv6Tree;
+import net.ripe.db.whois.common.rpsl.AttributeType;
+import net.ripe.db.whois.common.rpsl.ObjectType;
+import net.ripe.db.whois.common.rpsl.RpslObject;
+import net.ripe.db.whois.common.rpsl.attrs.Domain;
+import net.ripe.db.whois.common.rpsl.attrs.Inet6numStatus;
+import net.ripe.db.whois.common.rpsl.attrs.InetnumStatus;
+import org.eclipse.jetty.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Set;
+
+import static net.ripe.db.whois.common.rpsl.attrs.Inet6numStatus.ALLOCATED_BY_RIR;
+import static net.ripe.db.whois.common.rpsl.attrs.InetnumStatus.ALLOCATED_UNSPECIFIED;
+
+
+@Service
+public class RdapRelationService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RdapRelationService.class);
+
+    private final Ipv4Tree ip4Tree;
+    private final Ipv6Tree ip6Tree;
+    private final Ipv4DomainTree ipv4DomainTree;
+    private final Ipv6DomainTree ipv6DomainTree;
+    private final RpslObjectDao rpslObjectDao;
+
+    @Autowired
+    public RdapRelationService(final Ipv4Tree ip4Tree, final Ipv6Tree ip6Tree,
+                               final Ipv4DomainTree ipv4DomainTree, final Ipv6DomainTree ipv6DomainTree,
+                               final RpslObjectDao rpslObjectDao) {
+        this.ip4Tree = ip4Tree;
+        this.ip6Tree = ip6Tree;
+        this.ipv4DomainTree = ipv4DomainTree;
+        this.ipv6DomainTree = ipv6DomainTree;
+        this.rpslObjectDao = rpslObjectDao;
+    }
+
+    public List<String> getDomainRelationPkeys(final String pkey, final RelationType relationType){
+        final Domain domain = Domain.parse(pkey);
+        final IpInterval reverseIp = domain.getReverseIp();
+        final List<IpEntry> ipEntries = getIpEntries(getIpDomainTree(reverseIp), relationType, reverseIp);
+
+        return ipEntries
+                .stream()
+                .map(ipEntry -> rpslObjectDao.getById(ipEntry.getObjectId()).getKey().toString())
```

Comment:
why you need to hit the database? Can this throw exception if object is deleted? I think even if deleted pkey will be available

Reply:
Discussed over call
Here I have the IP and the Domain's object ID
I am using that object ID to fetch the reverse IP...We are not going to implement a logic to get the reverse IP out of the IP now so I added a TODO